sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -563,12 +563,29 @@ public static Unbound withSideInputs(
               DoFn.class.getSimpleName()));
     }
 
+    // To be removed when the features are complete and runners have their own adequate
+    // rejection logic
+    if (!signature.timerDeclarations().isEmpty()) {
+      throw new UnsupportedOperationException(
+          String.format("Found %s annotations on %s, but %s cannot yet be used with timers.",
+              DoFn.TimerId.class.getSimpleName(),
+              fn.getClass().getName(),
+              DoFn.class.getSimpleName()));
+    }
+
     // State is semantically incompatible with splitting
-    if (!signature.stateDeclarations().isEmpty()) {
+    if (!signature.stateDeclarations().isEmpty() && signature.processElement().isSplittable()) {
```

Comment:
The error messages are different. Nesting the secondary condition within `if (isSplittable)` increases the cyclomatic complexity for a reader without saving anything. Did I understand what you mean?